### DISCUSSION TITLE
Relay circuit-v2 expiration

### DISCUSSION THREAD

**Discussion by mkuhn19** [libp2p/rust-libp2p#4279] [Q&A] [answered]
Hi,

I am trying to develop a peer to peer application using rust and libp2p. I have nodes behind NAT and therefore need to use a mechanism such as relays to allow them to communicate. 

I just read the [relay circuit-v2 specs](https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md#circuit-relay-v2--).
As far as I understood, when a node makes a reservation request, the relay responds with a HopMessage. The message should contains the `reservation` field with an `expire` field inside, as described [there](https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md#reservation. ).

I have questions about this `expire` field :
1) I believe the node needs to make a new reservation request after `expire` time. Is that correct ?  

2) How can I read the `expire` field ?  
Here is how I proceed :  
Firstly, the node makes a reservation request by listening on the relay address with the P2pCircuit protocol.
```rust
let relay_addr = Multiaddr::empty()
                .with(Protocol::Ip4(Ipv4Addr::new(127,0,0,1)))
                .with(Protocol::Tcp(RELAY_PORT))
                .with(Protocol::P2p(relay_peer_id()))
                .with(Protocol::P2pCircuit);
            self.swarm.listen_on(relay_addr)?;
```
The relay_addr looks like /ip4/192.168.1.19/tcp/62500/p2p/12D3KooWMAXwrRcBdK3hFECY7b69PVW5rfHRa2WQPmbmMezZnEVG/p2p-circuit/ 
(When one wants to connect to the node, it uses the relay_addr and adds p2p/<NODE_PEER_ID>.). 

To handle the reservation request event, my code looks like this :
```rust
loop {
    event = self.swarm.select_next_some() => match event {
         SwarmEvent::Behaviour(ClientBehaviourEvent::Relay(relay::client::Event::ReservationReqAccepted { relay_peer_id, renewal, limit })) => {
                        // ....
                    }

    }
}
```
Shouldn't the `expire`field be there ?
I think I misunderstood something here, can you help me ?

Thank you !

**Comment by thomaseizinger** [comment, marked as answer]
The reservation will be automatically renewed in the background: https://github.com/libp2p/rust-libp2p/blob/c86b665a653626f5eca30c4f8811a2a08c3fbb20/protocols/relay/src/priv_client/handler.rs#L661-L667

**Reply by mkuhn19** [reply]
Thank you for such a fast answer 😄
Have a nice day !
